trusted_applet/internal/storage/README.md

```diff
@@ -0,0 +1,155 @@
+-------------|-----------------------------|-------------------------
+`Magic`      |`[4]byte{'T', 'F', 'J', '0'}`| Magic record header, v0
+`Revision`   |`uint32`                     | Incremented with each write to slot
+`DataLen`    |`uint64`                     | `len(RecordData)`
+`Checksum`   |`[32]byte{}`                 | `SHA256` of `RecordData`
+`RecordData` |`[DataLen]byte{}`            | Application data
+
+
+An update record is considered _valid_ if its:
+
+*   `Magic` is correct,
+*   `Checksum` is correct for the `RecordData`
+
+The first time `Open` is called for a given slot, the slot's journal will be scanned from the beginning to look for the valid update record with the largest `Revision`. The Data from this record is the data associated with the slot. It could potentially be cached in RAM at this point if it's small enough.
+
+If no such record exists, then the slot has not yet been successfully written to and there is no data associated with the slot.
+
+An update to the slot causes an update record to be written to the journal starting at either:
+
+*   The first byte of the sector following the extent of the "current" update record (i.e all sectors contain header/data for at most 1 record), if there is sufficient space remaining in the journal to accommodate the entire update record without wrapping around to the first sector, or
+*   The first byte of the first sector in the journal, if there is no current record or the update record will not fit in the remaining journal space.
+
+Following a successful write to storage, the metadata associated with slot (i.e. Revision, current header location, location for next write, etc.) is updated.
+
+The diagram below shows a sequence of several update record writes of varying data sizes. These writes are taking place in a single journal, which you'll remember comprises several sectors.
+
+The grey boxes represent sectors containing old/previous data, green represents sectors holding the latest successful write.
+
+The numbers indicate a header with a particular `Revision`, sectors with `…` contain follow-on `RecordData`, and an x indicates invalid record header:
+
+```
+⬛⬛⬛⬛⬛⬛⬛⬛⬛⬛ - Initial state, nothing written
+🟩🟩🟩⬛⬛⬛⬛⬛⬛⬛ - First record (rev=1) has been successfully stored
+⬜⬜⬜🟩🟩⬛⬛⬛⬛⬛ - Next record (rev=2) is stored with the next available block
+⬜⬜⬜⬜⬜🟩🟩🟩⬛⬛ - Same again.
+🟩🟩🟩⬜⬜⬜⬜⬜⬛⬛ - The 4th record will not fit in the remaining space, so is written starting at the zeroth sector, overwriting old revision(s) - note it does not wrap around.
+⬜⬜⬜🟩🟩🟩⬜⬜⬛⬛ - Subsequent revisions continue in this vein.
+```
+
+Since record revisions should always be increasing as we scan left-to-right through the slot storage, we can assume we've found the newest update record when we've either reached the end of the storage space, or after having read at least 1 _good_ update record we find a record with a lower `Revision` than the previous record, or one with an invalid `Magic` or `Checksum.`
```

Comment:
I don't think there's much benefit there: we're reading an MMC block at a time anyway (and this will probably be the bottleneck); but even if we read the revision before the magic, we'd have to read and check the magic before relying on it (e.g. imagine the "empty storage" - a Rev of [0x00, 0x00, 0x00, 0x00] is going to be a valid first rev, but wouldn't be a valid magic).